src/interface/meshblock_data.cpp

```diff
@@ -72,124 +82,129 @@ void MeshBlockData<T>::Add(const std::string &label, const Metadata &metadata) {
+        // otherwise we get an empty pool
+        const auto &sparse_pool = resolved_packages_->GetSparsePool(name);
+
+        // add all sparse ids of the pool
+        for (const auto iter : sparse_pool.pool()) {
+          // this variable must exist, if it doesn't something is very wrong
+          const auto &v = varMap_.at(MakeVarLabel(name, iter.first));
+          add_var(v);
+          found = true;
+        }
       }
-      found = true;
-      Add(fv->second);
-    }
-    if (!found) {
-      std::stringstream msg;
-      msg << "MeshBlockData: " << name << " found more than once!" << std::endl;
-      PARTHENON_THROW(msg);
+
+      PARTHENON_REQUIRE_THROWS(found, "MeshBlockData::CopyFrom: Variable '" + name +
+                                          "' not found");
     }
   }
 }
+
+// Constructor for getting sub-containers
+// the variables returned are all shallow copies of the src container.
+// Optionally extract only some of the sparse ids of src variable.
+template <typename T>
+MeshBlockData<T>::MeshBlockData(const MeshBlockData<T> &src,
+                                const std::vector<std::string> &names) {
+  CopyFrom(src, true, names);
+}
+
 template <typename T>
 MeshBlockData<T>::MeshBlockData(const MeshBlockData<T> &src,
                                 const std::vector<MetadataFlag> &flags) {
-  auto var_map = src.GetCellVariableMap();
-  auto sparse_map = src.GetSparseMap();
-  auto face_map = src.GetFaceMap();
-  for (auto &it : var_map) {
-    auto n = it.first;
-    auto v = it.second;
-    if (v->metadata().AnyFlagsSet(flags)) {
-      Add(v);
-    }
-  }
-  for (auto &it : sparse_map) {
-    auto n = it.first;
-    auto v = it.second;
-    if (v->metadata().AnyFlagsSet(flags)) {
-      Add(v);
-    }
-  }
-  for (auto &it : face_map) {
-    auto n = it.first;
-    auto v = it.second;
-    if (v->metadata().AnyFlagsSet(flags)) {
-      Add(v);
-    }
-  }
+  CopyFrom(src, true, {}, flags);
 }
 
 // provides a container that has a single sparse slice
 template <typename T>
-std::shared_ptr<MeshBlockData<T>> MeshBlockData<T>::SparseSlice(int id) {
+std::shared_ptr<MeshBlockData<T>> MeshBlockData<T>::SparseSlice(int sparse_id) {
   auto c = std::make_shared<MeshBlockData<T>>();
 
   // copy in private data
-  c->pmy_block = pmy_block;
+  c->SetBlockPointer(GetBlockPointer());
+  c->resolved_packages_ = resolved_packages_;
 
-  // Note that all standard arrays get added
-  // add standard arrays
-  for (auto v : varVector_) {
-    c->Add(v);
+  // Note that all dense variables get added
+  for (const auto &v : varVector_) {
+    if (!v->IsSparse() || (v->GetSparseID() == sparse_id)) {
+      c->Add(v);
+    }
   }
   // for (auto v : s->_edgeVector) {
   //   EdgeVariable *vNew = new EdgeVariable(v->label(), *v);
   //   c.s->_edgeVector.push_back(vNew);
   // }
-  for (auto v : faceVector_) {
+  for (const auto &v : faceVector_) {
```

Comment:
I just wondered if the error is caught we one tries to add a "sparse face field" right now (not that we're supporting that in any way).

Reply:
Yup, see lines 161 - 162 of `variable.hpp`:
```c++
assert(!metadata.IsSet(Metadata::Sparse) &&
       "Sparse not implemented yet for FaceVariable");
```
However, since that's just an `assert`, it only gets check in debug mode. I'll change it to a `PARTHENON_REQUIRE_THROWS`.